core/domain/classifier_services.py

```diff
@@ -0,0 +1,322 @@
+        # Always set default label
+        label_vector[self._label_to_id['_default']] = 1
+        return label_vector
+
+    def _update_counting_matrices(self, d, w, l, val):
+        """Update counting matrices given doc/word/label indices."""
+        self._c_dl[d, l] += val
+        self._c_lw[l, w] += val
+        self._c_l[l] += val
+
+    def _get_doc_ids(self, doc_ids):
+        if doc_ids is None:
+            doc_ids = xrange(self._doc_count)
+        return doc_ids
+
+    def _init_docs(self, doc_ids=None):
+        """Initialize data for given docs (defaults to all)."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            l_c = [
+                numpy.random.multinomial(1, labels / labels.sum()).argmax()
+                for i in xrange(len(doc))
+            ]
+            self._l_dc.append(l_c)
+            for w, l in zip(doc, l_c):
+                self._update_counting_matrices(d, w, l, 1)
+
+    def _infer_docs(self, doc_ids):
+        """Runs iterative inference on given docs (default all)."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        statez = {
+            'updates': 0,
+            'computes': 0
+        }
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            for c in xrange(len(doc)):
+                w = doc[c]
+                l = self._l_dc[d][c]
+
+                self._update_counting_matrices(d, w, l, -1)
+
+                # Gibbs update of labels
+                coeff_a = 1 / (
+                    self._c_dl[d].sum() + self._label_count * self._alpha)
+                coeff_b = 1 / (
+                    self._c_lw.sum(axis=1) + self._word_count * self._beta)
+                prob_l = (
+                    labels *
+                    coeff_a * (self._c_dl[d] + self._alpha) *
+                    coeff_b * (self._c_lw[:, w] + self._beta))
+                new_l = numpy.random.multinomial(
+                    1, prob_l / prob_l.sum()
+                    ).argmax()
+
+                statez['computes'] += 1
+                if l != new_l:
+                    statez['updates'] += 1
+
+                self._l_dc[d][c] = new_l
+                self._update_counting_matrices(d, w, new_l, 1)
+
+        return statez
+
+    def _get_probabilities(self, d):
+        """Returns the probability of a document having a label."""
+        probs_i = self._c_dl[d] + (self._b_dl[d] * self._alpha)
+        probs_i = probs_i / probs_i.sum(axis=0)[numpy.newaxis]
+        return probs_i
+
+    def _get_probabilities_with_label(self, d):
+        """Includes human readable labels with document probabilities."""
+        probs_i = self._get_probabilities(d)
+        probs_l = {}
+        for label, l in self._label_to_id.iteritems():
+            probs_l[label] = probs_i[l]
+        return probs_l
+
+    def _predict(self, d, threshold=0.5):
+        """Calculates prediction data for a doc"""
```

Comment:
Hm, I originally returned a complex type for debug purposes - simplifying (and documenting) it.